### PR TITLE
Extract URL de-slashing logic to UriCleaner Service and fix query parameter handling - update tests

### PR DESCRIPTION
**Pull request by ChristianOsswald**
### Background / Problem

The extension used a simple suffix check to determine if a URL needed to be de-slashed:
     `str_ends_with($linkResult->getUrl(), '/')`

  This check failed for any URLs containing query parameters or fragment identifiers.

  #### Example:

  For the URL:
   https://my-site.ddev.site/en/?sitemap=pages&type=1533906435

  • The Issue: Because the URL ends with the query parameter value ( f ) instead of a slash ( / ), the
  str_ends_with()  check returned  false .
  • The Consequence: The trailing slash after the  /en/  segment was never stripped, leaving the URL in an
  inconsistent state ( /en/?...  instead of  /en?... ).

  ### Solution

  This PR resolves issues with trailing slash removal on URLs that contain query parameters, fragments, or user
  credentials. It also refactors the de-slashing logic into a centralized utility class prevent code duplication.

  #### What is achieved:

  1. Query Parameter & Fragment Support:
  URLs like  /en/?sitemap=pages  or  /en/#anchor are now correctly de-slashed to  /en?sitemap=pages  and  /en#anchor 
  (previously they were skipped because the overall URL string did not end with  / ).
  2. Centralized Logic ( UriUtility ):
  Extracted the URL parsing and reconstruction logic to a reusable helper class:
  UriUtility::removeTrailingSlashFromPath() . This logic is now shared across:
      •  DeSlashCanonicalUrl 
      •  DeSlashHrefLangGenerator 
      •  DeSlashingPageLinkBuilder 
  3. Robust URL Reconstruction & Edge-Case Handling:
      • Credentials Preservation: Correctly retains user/password credentials (e.g.  user:pass@host ) if present in
      the URL.
      • Protocol-Relative URLs: Preserves leading double slashes (e.g.,  //example.com/en/  ->  //example.com/en ).
      • Preventing Redirect Chains: Uses  rtrim  to clean up multiple trailing slashes at once (e.g.,  /en//  ->  /en), preventing unnecessary double redirects (like  /en//  ->  /en/  ->  /en ).
      • Slash-Only Paths Guard: Ensures slash-only paths (like  /  or  /// ) remain unaffected and are preserved as-is.
      • Empty Strings: Handled safely via a dedicated guard clause.
  5. Test Coverage:
      • Introduced a PHPUnit test suite covering 15 distinct edge cases for URL de-slashing.
      • Verified that all functional tests (checking routing and 301 redirects) are passing successfully as well.

### CODE REVIEW

**Comment by bmack** [comment]
Hey,

thank you for the change. However, I try to avoid Utility classes for new code. Can you instead make a new class "UriCleaner" which can be injected (= non-static method) into the relevant code, so it could be stubbed for the SUT in the future.

**Comment by ChristianOsswald** [comment]
> Hey,
> 
> thank you for the change. However, I try to avoid Utility classes for new code. Can you instead make a new class "UriCleaner" which can be injected (= non-static method) into the relevant code, so it could be stubbed for the SUT in the future.

Thanks for the feedback! I completely agree, dependency injection makes this much cleaner.

I've just force-pushed an update that addresses your points: 

- Extracted the logic into a new, non-static UriCleaner service.
- Refactored all three event listeners (DeSlashCanonicalUrl ,  DeSlashHrefLangGenerator , DeSlashingPageLinkBuilder) to use constructor injection.
- Added unit tests for all the event listeners
- I noticed the GitHub Actions ci.yml wasn't actually configured to run the Unit Tests, so I added a step for that as well. 

Let me know if everything looks good to you now!

**Review by bmack — APPROVED** [review]

**Comment by bmack** [comment]
looks good @ChristianOsswald - tests fail though

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,15 +1,24 @@
...
 
-    runs-on: ubuntu-latest
+    name: all tests
+    runs-on: ubuntu-22.04
     strategy:
+      fail-fast: false
       matrix:
-        TYPO3: [ 12.4, 13.4]
-        PHP: [ 8.2, 8.3 ]
+        TYPO3: [ '13.4', '14.0' ]
```
Comment (achimfritz) [inline .github/workflows/ci.yml]: i suggest 14.3

File: composer.json [hunk 2]
```diff
@@ -5,7 +5,7 @@
 	"keywords": ["seo", "typo3", "trailing slash"],
 	"license": "GPL-2.0-or-later",
     "require": {
-        "typo3/cms-core": "^12.4 || ^13.4"
+        "typo3/cms-core": "^13.4 || ^14.0"
```
Comment (achimfritz) [inline composer.json:8]: s. above, we should use 14.3

File: Classes/Service/UriCleaner.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+     * Remove trailing slash from the path segment of a URL, preserving query parameters and fragments.
+     * Returns the modified URL, or the original if no trailing slash in path.
+     */
+    public function removeTrailingSlashFromPath(string $url): string
+    {
+        if ($url === '') {
+            return $url;
+        }
+
+        $parts = parse_url($url);
```
Comment (achimfritz) [inline Classes/Service/UriCleaner.php]: i would prefer to use  `TYPO3\CMS\Core\Http\Uri`

**Review by achimfritz — COMMENTED** [review]
nice PR, thanks, would you apply my suggestions?

**Comment by ChristianOsswald** [comment]
Yeah, sure. Thank you for the feedback.

I will also fix the workflow file: (Line: 19, Col: 7): Unexpected value 'exclude'

**Comment by achimfritz** [comment]
i will fix the pipeline, if neccessary
